Fixed #30973 -- Converted selenium tests wait_page_loaded() to context manager.

codingjoe opened the pull request:
https://code.djangoproject.com/ticket/30973

codingjoe commented:
buildbot, test on selenium.

carltongibson commented on django/contrib/admin/tests.py:
```diff
@@ -97,14 +99,34 @@ def wait_until_invisible(self, css_selector, timeout=10):
             # Wait for the next page to be loaded
-            self.wait_for('body')
+            self.wait_until(ec.staleness_of(old_page), timeout=timeout)
+            self.wait_until(
```
This is just `wait_page_ready()` no?

codingjoe replied:
Good point, that was just natural code progression. I wrote one before the other and didn't refactor at the end 🤷‍♂

codingjoe commented:
buildbot, test on selenium.

pope1ni commented on django/contrib/admin/tests.py:
```diff
@@ -97,14 +99,29 @@ def wait_until_invisible(self, css_selector, timeout=10):
         )
 
-    def wait_page_loaded(self):
+    def wait_page_ready(self, timeout=10):
```
Make `timeout` keyword-only?
```python
def wait_page_ready(self, *, timeout=10):
    ...
```
Was there any educated choice behind the choice of 10 seconds? Would 5 or event 3 be enough for most test cases?

codingjoe replied:
10 seconds is standard in all other methods on this class.

I don't want to make it a keyword only options, as it would be inconsistent. If we change it, we should do it for the whole class and in a separate ticket. That being said, this is not public API.

pope1ni commented on django/contrib/admin/tests.py:
```diff
@@ -97,14 +99,29 @@ def wait_until_invisible(self, css_selector, timeout=10):
+        )
+
+    @contextmanager
+    def wait_page_loaded(self, timeout=10):
```
Make `timeout` keyword-only?

pope1ni commented on django/contrib/admin/tests.py:
```diff
@@ -97,14 +99,29 @@ def wait_until_invisible(self, css_selector, timeout=10):
+        """
+        self.wait_until(
+            lambda driver: driver.execute_script('return document.readyState;') == 'complete',
+            timeout
```
Add a trailing comma.

codingjoe replied:
Good catch, done!

pope1ni commented on django/contrib/admin/tests.py:
```diff
@@ -97,14 +99,29 @@ def wait_until_invisible(self, css_selector, timeout=10):
-    def wait_page_loaded(self):
+    def wait_page_ready(self, timeout=10):
+        """
+        Block until page has started to load.
```
This docstring is the same as for `wait_page_loaded()`. Do we even need it in either case? Or perhaps the docstring should explain the difference between these two methods.

codingjoe replied:
Yes, good point. I gave it a shot, be I fear it could still be improved.

pope1ni commented on django/contrib/admin/tests.py:
```diff
@@ -97,14 +99,29 @@ def wait_until_invisible(self, css_selector, timeout=10):
+            self.wait_until(ec.staleness_of(old_page), timeout=timeout)
+            self.wait_page_ready(timeout=timeout)
         except TimeoutException:
             # IE7 occasionally returns an error "Internet Explorer cannot
```
Do we even need this anymore? IE7 is *long* gone.

It also swallows all the `TimeoutException`s that we're expecting to happen here with the use of `wait_until()`, etc? Wouldn't we rather know that it timed out and we knew that was the cause than falling through to an obscure failure in a later assertion?

codingjoe replied:
I don't know, but good questions: @carltongibson do we test on Jenkins against IE7?

carltongibson replied:
I removed the try...except... — Thanks both.

codingjoe commented:
buildbot, test on selenium.

codingjoe commented:
@pope1ni thanks for the review! I addressed your comments. Let me know if you have further questions or suggestions. It seems everyone is happy get our selenium suite more stable 😄

carltongibson commented:
buildbot, test on selenium.

carltongibson commented:
buildbot, test on selenium.
